onnxruntime/core/mlas/lib/jblas_gemm.cpp

```diff
@@ -0,0 +1,525 @@
+                                DataParams[i].ldc
+                            ),
+                            size
+                        );
+                    } else if (NTile == tAVX512_VNNI::NTILE && _cd->AVX512_VNNI()) {
+                        size = std::max(
+                            JblasSQ4GemmCompInt8WorkspaceSize<tAVX512_VNNI>(
+                                M, N, K, DataParams[i].A, DataParams[i].lda, kptr, DataParams[i].C,
+                                DataParams[i].ldc
+                            ),
+                            size
+                        );
+                    } else if (NTile == tAVX_VNNI::NTILE && _cd->AVX_VNNI()) {
+                        size = std::max(
+                            JblasSQ4GemmCompInt8WorkspaceSize<tAVX_VNNI>(
+                                M, N, K, DataParams[i].A, DataParams[i].lda, kptr, DataParams[i].C,
+                                DataParams[i].ldc
+                            ),
+                            size
+                        );
+                    }
+                }
+                if (CType == uint32_t(gemm::CompType::COMP_INT8_SS_INT32)) {
+                    if (NTile == tAMX_INT8_SS::NTILE && _cd->AMX_INT8()) {
+                        size = std::max(
+                            JblasSQ4GemmCompInt8WorkspaceSize<tAMX_INT8_SS>(
+                                M, N, K, DataParams[i].A, DataParams[i].lda, kptr, DataParams[i].C,
+                                DataParams[i].ldc
+                            ),
+                            size
+                        );
+                    }
+                }
+            }
+            delete ptr;
+        }
+    }
+    return size;
+}
+
+template <typename T>
+static size_t
+JblasQ4BuSize(int block_size, size_t N, size_t K, bool isAsym)
+{
+    static T launcher;
+    auto stor = launcher.mProB.createStorage(
+        N, K, block_size, JBLAS_DTYPE::S4_CLIP, JBLAS_DTYPE::F32, JBLAS_DTYPE::BF16, isAsym
+    );
+    // TODO(Yu) support more scale dtype
+    return stor.mSize;
+}
+
+size_t
+JblasQ4GemmPackBSize(size_t N, size_t K, size_t BlkSize, bool isAsym, MLAS_SQNBIT_COMPUTE_TYPE CompType)
+{
+    GetCPUDevice();
+    if (K % BlkSize != 0) {
+        return 0;
+    }
+    // from low precision to high precision
+    switch (CompType) {
+        case CompInt8:
+            if (_cd->AMX_INT8() && BlkSize % tAMX_INT8_SS::KTILE == 0) {
+                return JblasQ4BuSize<tLauncher_Int8_S4_F32F32<tAMX_INT8_SS>>(int(BlkSize), N, K, isAsym);
+            }
+            if (_cd->AVX512_VNNI() && BlkSize % tAVX512_VNNI::KTILE == 0) {
+                return JblasQ4BuSize<tLauncher_Int8_S4_F32F32<tAVX512_VNNI>>(int(BlkSize), N, K, isAsym);
+            }
+            if (_cd->AVX_VNNI() && BlkSize % tAVX_VNNI::KTILE == 0) {
+                return JblasQ4BuSize<tLauncher_Int8_S4_F32F32<tAVX_VNNI>>(int(BlkSize), N, K, isAsym);
+            }
+        case CompBf16:
```

Comment:
Currently, CompType supports only three values: CompUndef, CompFp32 and CompInt8. The `CompBf16` and `CompFp16` cases are unreachable. Consider using `ORT_THROW("unreachable");` in such cases.

Reply:
@yihonglyu  It's not unreachable, they fall through to CompFp32. The compute type is not the fixed one, it's the minimum accuracy level, so if you don't support it or the device can not run it, it can fall through to a higher accuracy level like CompFp32.

Reply:
related question - could you explain why CompFp16 is considered higher accuracy than CompBf16?

Reply:
@edgchen1  It's the suggestion from the INC algorithm team. They prefer more mantissa bits. We also have one accuracy failure case on the LLM model with CompBF16.